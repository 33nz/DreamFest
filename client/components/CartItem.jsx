import React from 'react'

const CartItem = (props) => {
  const { name, id, quantity } = props.item
  const displayQuantity = (quantity === 0) ? '' : quantity
  return (
    <tr>
      <td>{name}</td>
      <td>
        <input
          aria-label='quantity'
          className='update-input'
          value={displayQuantity}
          onChange={(e) => props.update(id, e.target.value)} />
      </td>
      <td><button aria-label='delete' onClick={() => props.deleteFromCart(id)}>
        <span className='fa fa-trash fa-2x' />
      </button></td>
    </tr>
  )
}

export default CartItem
