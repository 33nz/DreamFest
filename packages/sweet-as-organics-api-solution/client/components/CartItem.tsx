<<<<<<< HEAD:packages/sweet-as-organics-api-solution/client/components/CartItem.tsx
import type * as React from 'react'
import { useDispatch } from 'react-redux'
import { CartItemWithQuantity } from '../../common/interfaces'

=======
import { ChangeEvent } from 'react'
>>>>>>> main:packages/sweet-as-organics-api/client/components/CartItem.tsx
import { deleteFromCart, updateCart } from '../actions/cart'
import { useAppDispatch } from '../hooks'

interface Props {
<<<<<<< HEAD:packages/sweet-as-organics-api-solution/client/components/CartItem.tsx
  item: CartItemWithQuantity
=======
  item: {
    name: string
    id: number
    quantity: number
  }
>>>>>>> main:packages/sweet-as-organics-api/client/components/CartItem.tsx
}

function CartItem(props: Props) {
  const { name, id, quantity } = props.item
  const dispatch = useAppDispatch()

  function update(e: ChangeEvent<HTMLInputElement>) {
    const newQuantity = Number(e.currentTarget.value)
    const isValidQuantity = !isNaN(newQuantity)

<<<<<<< HEAD:packages/sweet-as-organics-api-solution/client/components/CartItem.tsx
  function update(e: React.ChangeEvent<HTMLInputElement>) {
    const newQuantity = parseInt(e.target.value)
    if (!newQuantity) {
=======
    if (isValidQuantity) {
>>>>>>> main:packages/sweet-as-organics-api/client/components/CartItem.tsx
      const updateInfo = { id, newQuantity }
      dispatch(updateCart(updateInfo))
    }
  }

  function deleteItem() {
    dispatch(deleteFromCart(id))
  }

  const displayQuantity = quantity === 0 ? '' : quantity
  return (
    <tr>
      <td>{name}</td>
      <td>
        <input
          aria-label="quantity"
          className="update-input"
          value={displayQuantity}
          onChange={update}
        />
      </td>
      <td>
        <button aria-label="delete" onClick={deleteItem}>
          <span className="fa fa-trash fa-2x" />
        </button>
      </td>
    </tr>
  )
}

export default CartItem
