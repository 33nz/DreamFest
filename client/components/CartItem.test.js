import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import CartItem from './CartItem'

test('renders cart item correctly, with quantity in input', async () => {
  const mockItem = { id: 1, quantity: 4, name: 'mock cart item' }

  const tableBody = document.createElement('tbody')
  const { asFragment } = render(<CartItem item={mockItem} />, {
    container: document.body.appendChild(tableBody)
  })

  const itemName = await screen.getAllByRole('cell')[0]
  const quantityInput = await screen.getByRole('textbox', { name: 'quantity' })
  const deleteButton = await screen.getByRole('button', { name: 'delete' })

  expect(itemName).toHaveTextContent('mock cart item')
  expect(quantityInput.value).toBe('4')
  expect(deleteButton).toBeInTheDocument()
  expect(asFragment()).toMatchSnapshot()
})

test('renders empty input if quantity is 0', async () => {
  const mockItem = { id: 2, quantity: 0, name: 'mock cart item 2' }

  const tableBody = document.createElement('tbody')
  const { asFragment } = render(<CartItem item={mockItem} />, {
    container: document.body.appendChild(tableBody)
  })

  const quantityInput = await screen.getByRole('textbox', { name: 'quantity' })

  expect(quantityInput.value).toBe('')
  expect(asFragment()).toMatchSnapshot()
})

test('calls deleteFromCart on delete button click ', async () => {
  const mockItem = { id: 1, quantity: 4, name: 'mock cart item' }
  const deleteFromCart = jest.fn()

  const tableBody = document.createElement('tbody')
  render(<CartItem item={mockItem} deleteFromCart={deleteFromCart} />, {
    container: document.body.appendChild(tableBody)
  })

  const deleteButton = await screen.getByRole('button', { name: 'delete' })
  fireEvent.click(deleteButton)

  expect(deleteFromCart).toHaveBeenCalledWith(mockItem.id)
})

test('calls update on quantity field input change', async () => {
  const mockItem = { id: 1, quantity: 4, name: 'mock cart item' }
  const update = jest.fn()

  const tableBody = document.createElement('tbody')
  render(<CartItem item={mockItem} update={update} />, {
    container: document.body.appendChild(tableBody)
  })

  const quantityInput = await screen.getByRole('textbox', { name: 'quantity' })
  fireEvent.change(quantityInput, { target: { value: '5' } })

  expect(update).toHaveBeenCalledWith(mockItem.id, '5')
})
