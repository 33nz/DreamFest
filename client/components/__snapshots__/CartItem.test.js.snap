// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders cart item correctly, with quantity in input 1`] = `
<DocumentFragment>
  mock cart item
  <input
    aria-label="quantity"
    class="update-input"
    value="4"
  />
  <button
    aria-label="delete"
  >
    <span
      class="fa fa-trash fa-2x"
    />
  </button>
</DocumentFragment>
`;

exports[`renders empty input if quantity is 0 1`] = `
<DocumentFragment>
  mock cart item 2
  <input
    aria-label="quantity"
    class="update-input"
    value=""
  />
  <button
    aria-label="delete"
  >
    <span
      class="fa fa-trash fa-2x"
    />
  </button>
</DocumentFragment>
`;
